PocketTutor-Team/pockettutor-app
# Add possibility to mark a tutor as favorite to be able to find it again easily

# Bookmark tutors to be able to find them later easily

### PR Description
The goal of the PR is to enable students to boomark tutors so that they can easily find them again later. You will see in a next section how this functionality is implemented. This PR aslo add test for the updated made.

### Bookmark a Tutor : How it works?
When a student choose a tutor (in the Tutor Matching phase) or when he has a lesson confirmed with a tutor, he will be able to click on the star at the top-right corner of the tutor details screen to mark the tutor as favorite. If a tutor is already selected as favorite, this will remove the tutor from the list of favorites.

Moreover, when we are in the overview page, as a student, we will be able to click on a star action button to see the list of our favorite tutors. We can click on whatever tutor in the list, which will display the tutor details and a button to ask a lesson with this specific tutor. Clicking on this button will direct us to the page to create a new lesson with all the fields to fill (title, description, date, location...).

Finally, when we are in the Tutor Matching phase (after the creation of a lesson), the list is filtered and sort as usual but favorite tutors are always displayed first in the screen, allowing us to find our favorite tutors very easily.

### Technical details
I have added a field `favoriteTutors` to the Profile class. This field is a list of tutor UIDs, that contain all the tutors the student has bookmarked as favorite.

There are on new screen to display the list of favorite tutors (`FavoriteTutors.kt`). To display the details of a favorite tutor, the file `SelectedTutorDetails.kt` is reused with an additionnal paramter `isFavoriteTutor` that is set to true. These files used components defined before to display the information about tutors.

Concerning the creation of a lesson from the favorite tutors page, it will direct us to the `AddLesson` screen. When all the fields will be completed, and the lesson confirmed, the lesson will be created with the status PENDING_TUTOR_CONFIRMATION and with the favorite tutor as tutorId. This implies that if the tutor refuse the lesson, it will be proposed to other tutors as usual and this is why we need to fill all fields during lesson creation (like the price for example).

### Tests contained in the PR
- `FavoriteTutorsTest.kt` and `SelectedTutorDetailsWithFavoriteTest.kt` => these two files add test for the new screen added to implement the new feature and the `SelectedTutorDetails` screen with the field `isFavoriteTutor` sets to true.
- `SelectedTutorDetailsTest.kt` => add a test to be sure the bookmark icon is displayed and that we can interact with it.
- `TutorMatchingTest.kt` => add a test to be sure the section containing the favorite tutors is displayed.
- `OverviewTest.kt` => add a test to be sure the bookmark icon is displayed and that it navigate to the correct screen.
- `AddLessonTest.kt` and `EditRequestedLessonTest.kt` => fix the tests by adding field name to avoid confusion when we add new field with default values (here I have added the field `favoriteTutors`).

### How to test the PR?
To test this PR, one could first run the tests I added, or directly run the MainActivity as a student, create a new lesson (or use another confirmed lesson), bookmark a tutor, go in the overview favorite tutor section and see if everything works as intended.

### Screenshots of the modified screen
1) In the Overview, we have at the top, a star icon that allows us to see the list of favorite tutors:
![image](https://github.com/user-attachments/assets/ebe549be-3900-4cb2-bbcd-b2d473db032b)

2) When we click on the star in the overview, we can see the list of the tutors we marked as favorite:
![image](https://github.com/user-attachments/assets/284d557a-ee9e-425d-83fe-dc7dbd453d17)

3) In the list of favorite tutors, when we click on a tutor, we can see his profile details with a button to directly ask a lesson to him:
![image](https://github.com/user-attachments/assets/5185ad3b-6df0-495e-a2b1-f847444f9ad9)

4) When we create a lesson and we are in the tutor matching phase, we can see our favorite tutors on top of the page:
![image](https://github.com/user-attachments/assets/4821f211-48a4-4444-aa49-0bb9cece477f)

5) When we see the details of a tutor, we can bookmark it in our favorite by clicking on the star in the top-right corner. This start is yellow when the tutor is in our favorite, dark otherwise:
![image](https://github.com/user-attachments/assets/254e0397-3e45-423d-8426-3daa7ad62510) ![image](https://github.com/user-attachments/assets/a691da9f-b031-4e3c-b881-a9c9e5941490)
